File: robot/ros2_ws/src/cpp_subscriber_pkg/src/listener_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <my_interfaces/msg/greeting.hpp>

class ListenerNode : public rclcpp::Node
{
public:
  ListenerNode() : Node("cpp_listener_node")
  {
    subscription_ = create_subscription<my_interfaces::msg::Greeting>(
      "greetings", 10,
      [this](const my_interfaces::msg::Greeting & msg) {
        RCLCPP_INFO(get_logger(), "Received from \"%s\" (#%d): %s",
          msg.sender.c_str(), msg.count, msg.message.c_str());
      });
    RCLCPP_INFO(get_logger(), "CppListenerNode started, subscribing to /greetings");
  }

private:
  rclcpp::Subscription<my_interfaces::msg::Greeting>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ListenerNode>());
  rclcpp::shutdown();
  return 0;
}

File: robot/ros2_ws/src/subscriber_pkg/subscriber_pkg/listener_node.py
import random
import rclpy
from rclpy.node import Node
from my_interfaces.msg import Greeting
from rich.pretty import pretty_repr


class ListenerNode(Node):
    def __init__(self):
        super().__init__('listener_node')
        self.subscription = self.create_subscription(
            Greeting,
            'greetings',
            self.on_greeting,
            10,
        )
        self.data = [0] * 10
        self.get_logger().info('ListenerNode started, subscribing to /greetings')

    def on_greeting(self, msg: Greeting):
        self.data[random.randint(0, len(self.data) - 1)] = msg.count
        self.get_logger().info(pretty_repr(self.data))

        # if msg.count == 3:
            # raise Exception('Simulated error after receiving 3 messages')


def main(args=None):
    rclpy.init(args=args)
    node = ListenerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()
